// react-email/dist/components/container/container.cjs
const require_runtime = require("../../_virtual/_rolldown/runtime.cjs");
const require_element_marker = require("../element-marker.cjs");
let react = require("react");
react = require_runtime.__toESM(react);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/container/container.tsx
const Container = react.forwardRef(({ children, style = {}, ...props }, ref) => {
	const tdStyle = {};
	const tableStyle = {};
	const styleRecord = style;
	for (const key in styleRecord) {
		if (!Object.hasOwn(styleRecord, key)) continue;
		const value = styleRecord[key];
		if (key === "padding" || key === "paddingTop" || key === "paddingRight" || key === "paddingBottom" || key === "paddingLeft") tdStyle[key] = value;
		else tableStyle[key] = value;
	}
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("table", {
		align: "center",
		width: "100%",
		...props,
		border: 0,
		cellPadding: "0",
		cellSpacing: "0",
		ref,
		role: "presentation",
		style: {
			maxWidth: "37.5em",
			...tableStyle
		},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("tbody", { children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("tr", {
			style: { width: "100%" },
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("td", {
				style: tdStyle,
				children
			})
		}) })
	});
});
Container.displayName = "Container";
require_element_marker.markAsElement(Container);
//#endregion
exports.Container = Container;
