// react-email/dist/components/container/container.mjs
import { markAsElement } from "../element-marker.mjs";
import * as React$1 from "react";
import { jsx } from "react/jsx-runtime";
//#region src/components/container/container.tsx
const Container = React$1.forwardRef(({ children, style = {}, ...props }, ref) => {
	const tdStyle = {};
	const tableStyle = {};
	const styleRecord = style;
	for (const key in styleRecord) {
		if (!Object.hasOwn(styleRecord, key)) continue;
		const value = styleRecord[key];
		if (key === "padding" || key === "paddingTop" || key === "paddingRight" || key === "paddingBottom" || key === "paddingLeft") tdStyle[key] = value;
		else tableStyle[key] = value;
	}
	return /* @__PURE__ */ jsx("table", {
		align: "center",
		width: "100%",
		...props,
		border: 0,
		cellPadding: "0",
		cellSpacing: "0",
		ref,
		role: "presentation",
		style: {
			maxWidth: "37.5em",
			...tableStyle
		},
		children: /* @__PURE__ */ jsx("tbody", { children: /* @__PURE__ */ jsx("tr", {
			style: { width: "100%" },
			children: /* @__PURE__ */ jsx("td", {
				style: tdStyle,
				children
			})
		}) })
	});
});
Container.displayName = "Container";
markAsElement(Container);
//#endregion
export { Container };
